File: frontend/src/shared/contracts/notice.ts
import type { TFunction } from "@/shared/i18n";

export type NoticeKind = "runtime_event" | "system_status" | "product_update";
export type NoticeCategory =
  | "heartbeat"
  | "cron"
  | "subagent"
  | "exec"
  | "gateway"
  | "update";
export type NoticeSeverity = "info" | "success" | "warning" | "error" | "critical";
export type NoticeStatus = "unread" | "read" | "archived";
export type NoticeSurface = "center" | "toast" | "popup" | "os" | "footer";

export interface NoticeSource {
  producer: string;
  sessionKey: string;
  threadId: string;
  runId: string;
  jobId: string;
  channel: string;
  metadata?: Record<string, string>;
}

export interface NoticeAction {
  type: string;
  labelKey: string;
  target: string;
  params?: Record<string, string>;
}

export interface NoticeI18n {
  titleKey: string;
  summaryKey: string;
  bodyKey: string;
  params?: Record<string, string>;
}

export interface Notice {
  id: string;
  kind: NoticeKind;
  category: NoticeCategory;
  code: string;
  severity: NoticeSeverity;
  status: NoticeStatus;
  i18n: NoticeI18n;
  source: NoticeSource;
  action: NoticeAction;
  surfaces: NoticeSurface[];
  dedupKey: string;
  occurrenceCount: number;
  metadata?: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
  lastOccurredAt: string;
  readAt?: string;
  archivedAt?: string;
  expiresAt?: string;
}

export interface NoticeListRequest {
  statuses?: NoticeStatus[];
  kinds?: NoticeKind[];
  categories?: NoticeCategory[];
  severities?: NoticeSeverity[];
  surface?: NoticeSurface;
  query?: string;
  limit?: number;
}

export const NOTICE_CENTER_SURFACE: NoticeSurface = "center";
export const NOTICE_RUNTIME_I18N_KEYS = [
  "notifications.center.codes.heartbeatPeriodic.title",
  "notifications.center.codes.heartbeatPeriodic.summary",
  "notifications.center.codes.heartbeatPeriodic.body",
  "notifications.center.codes.heartbeatEvent.title",
  "notifications.center.codes.heartbeatEvent.summary",
  "notifications.center.codes.heartbeatEvent.body",
  "notifications.center.codes.heartbeatCron.title",
  "notifications.center.codes.heartbeatCron.summary",
  "notifications.center.codes.heartbeatCron.body",
  "notifications.center.codes.heartbeatExec.title",
  "notifications.center.codes.heartbeatExec.summary",
  "notifications.center.codes.heartbeatExec.body",
  "notifications.center.codes.heartbeatSubagent.title",
  "notifications.center.codes.heartbeatSubagent.summary",
  "notifications.center.codes.heartbeatSubagent.body",
  "notifications.center.codes.heartbeatRuntimeFailed.title",
  "notifications.center.codes.heartbeatRuntimeFailed.summary",
  "notifications.center.codes.heartbeatRuntimeFailed.body",
  "notifications.footer.codes.appUpdate.title",
  "notifications.footer.codes.appUpdate.summary",
  "notifications.footer.codes.appUpdate.body",
  "notifications.footer.codes.externalToolsUpdate.title",
  "notifications.footer.codes.externalToolsUpdate.summary",
  "notifications.footer.codes.externalToolsUpdate.body",
  "notifications.actions.openThread",
  "notifications.actions.openCron",
  "notifications.actions.openNotifications",
  "notifications.actions.openAppUpdates",
  "notifications.actions.openExternalTools",
] as const;

export function noticeSeverityToIntent(severity: NoticeSeverity): "info" | "success" | "warning" | "danger" {
  switch (severity) {
    case "success":
      return "success";
    case "warning":
      return "warning";
    case "error":
    case "critical":
      return "danger";
    default:
      return "info";
  }
}

export function formatNoticeText(template: string, params?: Record<string, string>): string {
  if (!params) {
    return template;
  }
  let output = template;
  Object.entries(params).forEach(([key, value]) => {
    output = output.split(`{${key}}`).join(value ?? "");
  });
  return output;
}

export function resolveNoticeTitle(notice: Notice, t: TFunction): string {
  return formatNoticeText(t(notice.i18n.titleKey), notice.i18n.params);
}

export function resolveNoticeSummary(notice: Notice, t: TFunction): string {
  return formatNoticeText(t(notice.i18n.summaryKey), notice.i18n.params);
}

export function resolveNoticeBody(notice: Notice, t: TFunction): string {
  return formatNoticeText(t(notice.i18n.bodyKey), notice.i18n.params);
}

File: internal/domain/notice/notice.go
package notice

import (
	"context"
	"errors"
	"time"
)

var ErrNoticeNotFound = errors.New("notice not found")

type Kind string

const (
	KindRuntimeEvent  Kind = "runtime_event"
	KindSystemStatus  Kind = "system_status"
	KindProductUpdate Kind = "product_update"
)

type Category string

const (
	CategoryHeartbeat Category = "heartbeat"
	CategoryCron      Category = "cron"
	CategorySubagent  Category = "subagent"
	CategoryExec      Category = "exec"
	CategoryGateway   Category = "gateway"
	CategoryUpdate    Category = "update"
)

type Severity string

const (
	SeverityInfo     Severity = "info"
	SeveritySuccess  Severity = "success"
	SeverityWarning  Severity = "warning"
	SeverityError    Severity = "error"
	SeverityCritical Severity = "critical"
)

type Status string

const (
	StatusUnread   Status = "unread"
	StatusRead     Status = "read"
	StatusArchived Status = "archived"
)

type Surface string

const (
	SurfaceCenter Surface = "center"
	SurfaceToast  Surface = "toast"
	SurfacePopup  Surface = "popup"
	SurfaceOS     Surface = "os"
	SurfaceFooter Surface = "footer"
)

type Source struct {
	Producer   string            `json:"producer"`
	SessionKey string            `json:"sessionKey"`
	ThreadID   string            `json:"threadId"`
	RunID      string            `json:"runId"`
	JobID      string            `json:"jobId"`
	Channel    string            `json:"channel"`
	Metadata   map[string]string `json:"metadata,omitempty"`
}

type Action struct {
	Type     string            `json:"type"`
	LabelKey string            `json:"labelKey"`
	Target   string            `json:"target"`
	Params   map[string]string `json:"params,omitempty"`
}

type I18n struct {
	TitleKey   string            `json:"titleKey"`
	SummaryKey string            `json:"summaryKey"`
	BodyKey    string            `json:"bodyKey"`
	Params     map[string]string `json:"params,omitempty"`
}

type Notice struct {
	ID              string         `json:"id"`
	Kind            Kind           `json:"kind"`
	Category        Category       `json:"category"`
	Code            string         `json:"code"`
	Severity        Severity       `json:"severity"`
	Status          Status         `json:"status"`
	I18n            I18n           `json:"i18n"`
	Source          Source         `json:"source"`
	Action          Action         `json:"action"`
	Surfaces        []Surface      `json:"surfaces"`
	DedupKey        string         `json:"dedupKey"`
	OccurrenceCount int            `json:"occurrenceCount"`
	Metadata        map[string]any `json:"metadata,omitempty"`
	CreatedAt       time.Time      `json:"createdAt"`
	UpdatedAt       time.Time      `json:"updatedAt"`
	LastOccurredAt  time.Time      `json:"lastOccurredAt"`
	ReadAt          *time.Time     `json:"readAt,omitempty"`
	ArchivedAt      *time.Time     `json:"archivedAt,omitempty"`
	ExpiresAt       *time.Time     `json:"expiresAt,omitempty"`
}

type ListFilter struct {
	Statuses   []Status
	Kinds      []Kind
	Categories []Category
	Severities []Severity
	Surface    Surface
	Query      string
	Limit      int
}

type Store interface {
	Save(ctx context.Context, item Notice) error
	Get(ctx context.Context, id string) (Notice, error)
	List(ctx context.Context, filter ListFilter) ([]Notice, error)
	CountUnread(ctx context.Context, surface Surface) (int, error)
	FindByDedupKey(ctx context.Context, dedupKey string) (Notice, error)
	MarkRead(ctx context.Context, ids []string, read bool, at time.Time) error
	Archive(ctx context.Context, ids []string, archived bool, at time.Time) error
	MarkAllRead(ctx context.Context, surface Surface, at time.Time) error
}
